### PR TITLE
Fixup/load balance get or create

### PR DESCRIPTION
**Pull request by rejoshed**
*Issue #, if available:* N/A

*Description of changes:*
Bug fix for Isolated network case.

*Testing performed:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/cloudstackmachine_controller.go [hunk 1]
```diff
@@ -137,9 +137,14 @@ func (r *CloudStackMachineReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		}
 	}
 
+	if !csCluster.Status.Ready {
```
Comment (rejoshed) [inline controllers/cloudstackmachine_controller.go:140]: Good prereq for all machines. May eventually move this to a watch filter.

File: pkg/cloud/network.go [hunk 2]
```diff
@@ -132,14 +133,14 @@ func (c *client) CreateIsolatedNetwork(csCluster *capcv1.CloudStackCluster) (ret
 	setIfNotEmpty(csCluster.Status.DomainID, p.SetDomainid)
 	resp, err := c.cs.Network.CreateNetwork(p)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "error encountered when creating network with name: %s", netStatus.Name)
```
Comment (rejoshed) [inline pkg/cloud/network.go:136]: Putting wrapfs into place in a lot of places as the Cloudstack error messages aren't really enough context.

File: pkg/cloud/network.go [hunk 3]
```diff
@@ -132,14 +133,14 @@ func (c *client) CreateIsolatedNetwork(csCluster *capcv1.CloudStackCluster) (ret
...
 	}
 
 	// Update Zone/Network status accordingly.
 	netStatus.ID = resp.Id
 	netStatus.Type = resp.Type
 	zoneStatus.Network = netStatus
-	csCluster.Status.Zones[zoneStatus.Name] = zoneStatus
+	csCluster.Status.Zones[zoneStatus.ID] = zoneStatus
```
Comment (rejoshed) [inline pkg/cloud/network.go:143]: Yeah, this wasn't good.

File: pkg/cloud/network.go [hunk 4]
```diff
@@ -234,25 +233,37 @@ func (c *client) FetchPublicIP(csCluster *capcv1.CloudStackCluster) (*cloudstack
...
-		return err
+		return errors.Wrapf(err, "error encountered while fetching a public IP address")
 	}
 
 	csCluster.Spec.ControlPlaneEndpoint.Host = publicAddress.Ipaddress
 	csCluster.Status.PublicIPID = publicAddress.Id
 
+	// Check if the address is already associated with the network.
 	zoneStatus := csCluster.Status.Zones.GetOne()
+	if publicAddress.Associatednetworkid == zoneStatus.Network.ID {
```
Comment (rejoshed) [inline pkg/cloud/network.go:244]: Here's the main fix. Don't try to re-associate the network if it's already associated.

File: pkg/cloud/network.go [hunk 5]
```diff
@@ -343,9 +354,9 @@ func (c *client) GetOrCreateLoadBalancerRule(csCluster *capcv1.CloudStackCluster
 	return nil
 }
 
-func (c *client) DestroyNetwork(net capcv1.Network) (retErr error) {
-	_, retErr = c.cs.Network.DeleteNetwork(c.cs.Network.NewDeleteNetworkParams(net.ID))
-	return retErr
+func (c *client) DeleteNetwork(net capcv1.Network) error {
```
Comment (rejoshed) [inline pkg/cloud/network.go:357]: Everything else is delete, so I made this delete too.

File: pkg/cloud/network_test.go [hunk 6]
```diff
@@ -276,4 +276,41 @@ var _ = Describe("Network", func() {
 			Ω(client.ResolveNetwork(dummies.CSCluster, &dummies.ISONet1)).Should(Succeed())
 		})
 	})
+
+	Context("Network Semi-Integ Tests", func() {
```
Comment (rejoshed) [inline pkg/cloud/network_test.go:280]: New integ test that could have prevented the whole fiasco. So, at least we shouldn't fail this way again.

File: test/dummies/vars.go [hunk 7]
```diff
@@ -171,7 +171,7 @@ func SetDummyCAPCClusterVars() {
 	CSClusterKind = "CloudStackCluster"
 	CSClusterName = "test-cluster"
 	EndPointHost = "EndpointHost"
-	EndPointPort = int32(8675309)
+	EndPointPort = int32(5309)
```
Comment (rejoshed) [inline test/dummies/vars.go:174]: TBH, kidna sad that 7 digits is too much for a real port here. I _do_ like my easter eggs.

**Review by maxdrib — APPROVED** [review]
I don't fully understand all the implications of these changes, but if you say it fixes the problem reported and it improves error logging, then I'm happy :)

**Pull request merged**